pkg/loopback/loopback_test.go

```diff
@@ -0,0 +1,136 @@
+// +build linux,cgo
+
+package loopback
+
+import (
+	"flag"
+	"io/ioutil"
+	"os"
+	"strings"
+	"syscall"
+	"testing"
+
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const filePath = "/tmp/"
+
+var rootEnabled bool
+
+func init() {
+	flag.BoolVar(&rootEnabled, "test.root", false, "enable tests that require root")
+}
+
+func RequiresRoot(t *testing.T) {
+	if !rootEnabled {
+		t.Skip("skipping test that requires root")
+		return
+	}
+	require.Equal(t, 0, os.Getuid(), "This test must be run as root.")
+}
+
+func createSparseFile() (string, error) {
+	size := int64(10 * 1024 * 1024)
+	fd, err := ioutil.TempFile(filePath, "sparseFile")
+	if err != nil {
+		return "", errors.Wrap(err, "Error creating file")
+	}
+	_, err = fd.Seek(size-1, 0)
+	if err != nil {
+		return "", errors.Wrap(err, "Failed to seek")
+	}
+	_, err = fd.Write([]byte{0})
+	if err != nil {
+		return "", errors.Wrap(err, "Write failed")
+	}
+	err = fd.Close()
+	if err != nil {
+		return "", errors.Wrap(err, "Close failed")
+	}
+	return fd.Name(), nil
+}
+
+func TestAttachLoopDevice(t *testing.T) {
+	RequiresRoot(t)
+	fd, err := AttachLoopDevice(filePath + "test.log")
+	require.Equal(t, ErrAttachLoopbackDevice, err)
+
+	fileName, err := createSparseFile()
+	require.NoError(t, err, "Error creating sparse file")
+	defer os.Remove(fileName)
+	fd, err = os.Open(fileName)
+	defer fd.Close()
+	require.NoError(t, err, "Error opening file")
+
+	loopDevice, err := AttachLoopDevice(fileName)
+	require.NoError(t, err, "Error attaching loop device")
+
+	if !strings.Contains(loopDevice.Name(), "/dev/loop") {
+		t.Fatal("Invalid loop file")
+	}
+
+	stat, err := fd.Stat()
+
+	targetINode := stat.Sys().(*syscall.Stat_t).Ino
+	targetDevice := stat.Sys().(*syscall.Stat_t).Dev
+	loopInfo, err := ioctlLoopGetStatus64(loopDevice.Fd())
+	require.NoError(t, err, "Error get loopback backing file")
+
+	if loopInfo.loDevice != targetDevice || loopInfo.loInode != targetINode {
+		t.Fatal("File not properly attached")
+	}
+}
+
+func TestFindLoopDeviceFor(t *testing.T) {
+	RequiresRoot(t)
+	fileName, err := createSparseFile()
+	require.NoError(t, err, "Error creating sparse file")
+	defer os.Remove(fileName)
+
+	fd, err := os.Open(fileName)
+	require.NoError(t, err, "Error opening file")
+	defer fd.Close()
+
+	loopDevice, err := AttachLoopDevice(fileName)
+	require.NoError(t, err, "Error attaching loop device")
+
+	confirmLoopDevice := FindLoopDeviceFor(fd)
```

Comment:
@dnephin the function `FindLoopDeviceFor` returns `nil` if we don't create a file inside a mounted volume. Also the targetINode and targetDevice present in line 77 and 78 respectively return different values while the file is not present in a mounted directory, hence the test fails. So for this we have to mount a volume while executing docker run.